A port of TroZ's NameColorFriendFoe script to ChromeShack

TroZ opened the pull request:
Adds a option to color usernames based upon the db user id. This makes it easier to pick out posts from the same user in a long thread, as well as makes the chatty more colorful
Option to change the name coloring algorithm between 4 options.
Adds more groups to the highlight users option (all new groups are disabled by default)
Allows user groups to add an emoji icon after user's name for groups of users.
Adds a menu item to the end of the user popup menu to easily add or remove a users from a highlight group without manually looking up their id.
TODO: need to figure out a good way to add new highlight groups for users that already have chrome shack installed.

electroly commented:
We already have this feature? It's "Highlight Users" and it has the same downside of requiring a user ID instead of a name. What I want instead is this: https://github.com/latestchatty/chromeshack/issues/97

TroZ commented:
Part of this change set is an option to give each user a unique color,
which is different from "Highlight Users".

The other part of the change set is an enhancement to "Highlight Users".
It adds more groups (but the new groups are all disabled to begin with),
but also adds to the popup menu so that you can see which groups that
person is in, and easily add or remove them from a group by clicking on the
group.
[image: image.png]
Here you can sew wft242 is in the 'star' group, and can remove him from
that group by clicking the star in the menu, or add him to other groups by
clicking their group names.

You can also see the per user name coloring that I added.


Adding by name in the option menu would be a good feature, but it would
involve somehow either looking up names from ids (I don't know is such an
API exists on Shacknews), or over time building up such a database on the
client side by scrapping the chatty.  This is nearly as useful, but much
easier to implement.

On Sat, Jun 8, 2019 at 12:03 AM Brian Luft <notifications@github.com> wrote:

> We already have this feature? It's "Highlight Users" and it has the same
> downside of requiring a user ID instead of a name. What I want instead is
> this: #97 <https://github.com/latestchatty/chromeshack/issues/97>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/latestchatty/chromeshack/pull/134?email_source=notifications&email_token=AAC22I3X2BUDJXIW37GVJMTPZMVP3A5CNFSM4HWADO32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXHMXFQ#issuecomment-500091798>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC22IYBIJ7PZW23BM7K573PZMVP3ANCNFSM4HWADO3Q>
> .
>

electroly commented:
Your images didn't come across. I recommend replying on the GitHub site.

By definition, you only care about usernames that appear on the page, and every username on the page has an id right there on the page too. Implementing that doesn't require persistent storage or doing anything over time. Realistically I don't think there's value in making any fancy feature improvements if it still requires you to enter IDs; people aren't going to do that (and they don't -- that feature has one foot in the grave). It needs to accept names if you want people to use it.

TroZ commented:
![image](https://user-images.githubusercontent.com/372003/59141943-a5fc3080-8983-11e9-95d3-6bb993757fd7.png)

What I have implemented does not require the user to enter any numbers.  User ID numbers are still shown in the Chrome Shack settings page, though.  Since this page can be opened outside of the chatty page, this is why I said that names would have to be scraped / stored or some Shack API for looking up names / ids, especially if you are on the options page and want the user to be able to enter a new name of a Shacker, who may not currently have posted for several days.

The way I have implemented it, you can open any post, click on the username in the full post, and add or remove that user from any of the non built in Highlight User groups.

WombatFromHell commented:
I've taken a bit of time to evaluate the code in this PR, and I can definitively say that it isn't going to be compatible with the changes coming in 1.64. It would need to be refactored in several ways:

1) I've moved the highlighted users arrays into `settings.js`, and rewritten both the Highlighted Users and Custom User Filter scripts to expose 2 methods: one to resolve users and their author ids on the current rendered page, and another to resolve the first instance of given user. I extensively performance tested several approaches when designing this, and I found that parsing the HTML of a `div.fullpost` (similar to what the `MutationObserver` does to raise events) was the most efficient way, and likewise doesn't require caching on most modern hardware. As a result, this PR would need to be refactored to call those resolvers rather than cludging together yet another resolver method that's likely to just slow page processing even more than we already do.

2) Due to some of the changes to the userpopup script seen in this PR I'm evaluating ways to make the current userpopup script more extensible to external scripts. Perhaps a set of deferred handlers that get called when userpopup first loads that let scripts add entries and callbacks to all instances of the menu?

In summary, this PR will probably need to wait for the changes to the settings system and user scripts to hit `master` before reevaluating it.

WombatFromHell commented:
This is just a quick follow-up to my previous comment: I believe the current implementation of Highlight Users on `master` approximates the behavior of this PR. I'd appreciate it if you could test against it and let me know if you'd like to suggest improvements, @TroZ.

TroZ commented:
I've had some computer issues but I have them resolved now (by building a
new computer).  I'll take a look at the changes this weekend and give you
my feedback.

On Wed, Aug 7, 2019 at 5:50 PM WombatFromHell <notifications@github.com>
wrote:

> This is just a quick follow-up to my previous comment: I believe the
> current implementation of Highlight Users on master approximates the
> behavior of this PR. I'd appreciate it if you could test against it and let
> me know if you'd like to suggest improvements, @TroZ
> <https://github.com/TroZ>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/latestchatty/chromeshack/pull/134?email_source=notifications&email_token=AAC22I73XTEP5GLCEKX3GW3QDM7SNA5CNFSM4HWADO32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3ZZ3YI#issuecomment-519282145>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC22I224BOF2RQZLDUOI3DQDM7SNANCNFSM4HWADO3Q>
> .
>

TroZ commented:
Wow, for the actual adding an removing Shackers from the various user
lists, this is much better than what I had implemented.  I like this a lot!

This does miss some of the other features of the change I submitted, but
this makes the highlight user feature much easier to use for the average
person.

My change included two other features, based off of my old 'Highlight User
Friend Foe' script.  The first, and a bit out of scope for the Chromeshack
user group highlighting, was an option color every user's name based upon
their user id, giving each user a color making it easier at a glance to see
where in a thread they posted. I had four coloring options that you could
choose from. I understand leaving this out for now.

The second feature is more inline with the chromeshack user highlighting
does.  I included a change that if the highlight user group name was a
single emoji character, instead of allowing a style to be set for the users
in that group, the emoji character would be put into a ::after content
where ever their name appeared.  This allowed you to easily save
'emotional' notes by putting 🙄 or 😡 or 💩 after user's names, or whatever
other emoji you wanted. These would appear directly after the user's name,
before the 'flair' badges.  Hopefully this can be re-implemented later.

On Wed, Aug 7, 2019 at 11:41 PM Brian Risinger <brian.risinger@gmail.com>
wrote:

> I've had some computer issues but I have them resolved now (by building a
> new computer).  I'll take a look at the changes this weekend and give you
> my feedback.
>
> On Wed, Aug 7, 2019 at 5:50 PM WombatFromHell <notifications@github.com>
> wrote:
>
>> This is just a quick follow-up to my previous comment: I believe the
>> current implementation of Highlight Users on master approximates the
>> behavior of this PR. I'd appreciate it if you could test against it and let
>> me know if you'd like to suggest improvements, @TroZ
>> <https://github.com/TroZ>.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/latestchatty/chromeshack/pull/134?email_source=notifications&email_token=AAC22I73XTEP5GLCEKX3GW3QDM7SNA5CNFSM4HWADO32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3ZZ3YI#issuecomment-519282145>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAC22I224BOF2RQZLDUOI3DQDM7SNANCNFSM4HWADO3Q>
>> .
>>
>

WombatFromHell commented:
Hmm, I'm not sure about how to implement that first suggestion. You've got a group that has a defined style assigned by default (a random color for all users in this new group as per the current implementation) and you also assign a random/unique color to each user in that group upon rendering I suppose? That seems like the inverse of what Highlight Users was originally designed to do.

I had in mind to revamp the style box to be more user friendly which would enable functionality similar to what you're describing but it would be very difficult in vanilla es6.

As for the second suggestion I could add that to my todo list as an enhancement in a future version.

TroZ commented:
The coloring all user's names is basically different from what highlight
users does.  It was basically a separate feature that upon page load would
get all the user names on the page, and add a line of css per user to color
their name.  It is implemented in my change about line 240
of highlight_users.js . Along with that, there was a button that would
modify the styles of the default highlight user groups to change the css of
the default groups from coloring the names, to putting a colored border
around the user names (as now the names would be 'randomly' colored). (Not
actually random, it was based upon the user id, which never changes for a
particular user).

I can add this feature back in later with a new pull request.  I think it
is better to get in your improvement to adding users to groups by name, and
adding and removing users using the popup menu when clicking on their name.

On Fri, Aug 9, 2019 at 7:29 PM WombatFromHell <notifications@github.com>
wrote:

> Hmm, I'm not sure about how to implement that first suggestion. You've got
> a group that has a defined style assigned by default (a random color for
> all users in this new group as per the current implementation) and you also
> assign a random/unique color to each user in that group upon rendering I
> suppose? That seems like the inverse of what Highlight Users was originally
> designed to do.
>
> I had in mind to revamp the style box to be more user friendly which would
> enable functionality similar to what you're describing but it would be very
> difficult in vanilla es6.
>
> As for the second suggestion I could add that to my todo list as an
> enhancement in a future version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/latestchatty/chromeshack/pull/134?email_source=notifications&email_token=AAC22I3KXPWDC64G4QUHAMDQDX4VBA5CNFSM4HWADO32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AAITA#issuecomment-520094796>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC22I7S3DKCBV6HSKPIZMTQDX4VBANCNFSM4HWADO3Q>
> .
>

WombatFromHell commented:
Chrome Shack 1.64 has been shipped with the reimplementation of this PR incorporated and credit given in the release notes. If you'd like to do a separate PR later on with additional enhancements rebased off the current `master` branch feel free to do so.
